servers/zts/src/main/java/com/yahoo/athenz/zts/store/DataStore.java

```diff
@@ -897,22 +897,27 @@ void processDomainRoles(DomainData domainData, DataCache domainCache) {
         }
 
         // Determine which roles have been deleted and should be removed from cache
-        final Set<String> validRoles = (roles != null) ? roles.stream().map(Role::getName).collect(Collectors.toSet()) : new HashSet<>();
+        // first we're going to extract our original roles. if we don't have
+        // any, then there is nothing to process, so we can return away
 
-        List<Role> deletedRoles = null;
         DataCache dataCache = getCacheStore().getIfPresent(domainData.getName());
-        if (dataCache != null) {
-            deletedRoles = dataCache.getDomainData().getRoles();
-            if (deletedRoles != null && !validRoles.isEmpty()) {
-                deletedRoles.removeIf(item -> validRoles.contains(item.getName()));
```

Comment:
here is what the issue is. deleteRoles is pointing to our cache so we cannot modify that list